vowpalwabbit/namespaced_feature_store.h

```diff
@@ -4,253 +4,217 @@
+  namespaced_feature_store& operator=(namespaced_feature_store&& other) = default;
+
+  inline size_t size() const
+  {
+    size_t accumulator = 0;
+    for (auto ns_index : _legacy_indices_existing) { accumulator += _feature_groups[ns_index].size(); }
+    return accumulator;
+  }
+  inline bool empty() const { return _legacy_indices_existing.empty(); }
 
   // Returns nullptr if not found.
-  features* get_feature_group(uint64_t hash);
-  // Returns nullptr if not found.
-  const features* get_feature_group(uint64_t hash) const;
+  inline features* get_or_null(namespace_index ns_index, uint64_t hash)
+  {
+    auto& bucket = _feature_groups[ns_index];
+    auto it = std::find_if(
+        bucket.begin(), bucket.end(), [hash](const namespaced_features& group) { return group.hash == hash; });
+    if (it == bucket.end()) { return nullptr; }
+    return &it->feats;
+  }
 
-  const std::set<namespace_index>& get_indices() const;
-  namespace_index get_index_for_hash(uint64_t hash) const;
+  // Returns nullptr if not found.
+  inline const features* get_or_null(namespace_index ns_index, uint64_t hash) const
+  {
+    auto& bucket = _feature_groups[ns_index];
+    auto it = std::find_if(
+        bucket.begin(), bucket.end(), [hash](const namespaced_features& group) { return group.hash == hash; });
+    if (it == bucket.end()) { return nullptr; }
+    return &it->feats;
+  }
 
-  // The following are experimental and may be superseded with namespace_index_begin_proxy
-  // Returns empty range if not found
-  std::pair<indexed_iterator, indexed_iterator> get_namespace_index_groups(namespace_index ns_index);
-  // Returns empty range if not found
-  std::pair<const_indexed_iterator, const_indexed_iterator> get_namespace_index_groups(namespace_index ns_index) const;
+  inline features& get(namespace_index ns_index, uint64_t hash)
+  {
+    auto* fs = get_or_null(ns_index, hash);
+    assert(fs != nullptr);
+    return *fs;
+  }
 
-  // If a feature group already exists in this "slot" it will be merged
-  template <typename FeaturesT>
-  features& merge_feature_group(FeaturesT&& ftrs, uint64_t hash, namespace_index ns_index);
+  inline const features& get(namespace_index ns_index, uint64_t hash) const
+  {
+    auto* fs = get_or_null(ns_index, hash);
+    assert(fs != nullptr);
+    return *fs;
+  }
 
   // If no feature group already exists a default one will be created.
   // Creating new feature groups will invalidate any pointers or references held.
-  features& get_or_create_feature_group(uint64_t hash, namespace_index ns_index);
+  features& get_or_create(namespace_index ns_index, uint64_t hash);
 
-#ifndef VW_NOEXCEPT
-  // These will throw if the hash does not exist
-  const features& operator[](uint64_t hash) const;
-  features& operator[](uint64_t hash);
-#endif
+  // UB if the list itself is modified.
+  inline std::list<namespaced_features>& get_list(namespace_index index) { return _feature_groups[index]; }
```

Comment:
can the list ever be modified by a different thread while we are holding a reference to it from here?

Reply:
No it shouldnt be because once parsing is done and it is passed to the learner that parse thread will never touch it again

Reply:
ok so this is more of a note to the developer

Reply:
Yep